api/app.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+#==============================================================================
+# Copyright (C) 2021-present Alces Flight Ltd.
+#
+# This file is part of Flight Job Service.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Job Service is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job Service. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job Service, please visit:
+# https://github.com/openflighthpc/flight-job-service
+#==============================================================================
+
+require 'sinatra/base'
+require 'sinatra/jsonapi'
+require_relative 'app/autoload'
+
+# Adds 401 Unauthorized error
+module Sinja
+  class UnauthorizedError < HttpError
+    HTTP_STATUS = 401
+
+    def initialize(*args) super(HTTP_STATUS, *args) end
+  end
+end
+
+# The base JSON:API for most interactions. Mounted in rack under
+# /:version
+class App < Sinatra::Base
+  register Sinatra::JSONAPI
+
+  helpers do
+    def role
+      case FlightJobAPI::PamAuth.valid?(env['HTTP_AUTHORIZATION'])
+      when true
+        :user
+      when false
+        :forbidden
+      else
+        raise Sinja::UnauthorizedError, 'Could not authenticate your authorization credentials'
+      end
+    end
+  end
+
+  configure_jsonapi do |c|
+    # Resource roles
+    c.default_roles = {
+      index: :user,
+      show: :user,
+      create: :user,
+      update: :user,
+      destroy: :user
+    }
+
+    # To-one relationship roles
+    c.default_has_one_roles = {
+      pluck: :user,
+      prune: :user,
+      graft: :user
+    }
+
+    # To-many relationship roles
+    c.default_has_many_roles = {
+      fetch: :user,
+      clear: :user,
+      replace: :user,
+      merge: :user,
+      subtract: :user
+    }
+  end
+
+  resource :authenticates, pkre: /user/ do
+    helpers do
+      def find(id)
+        Authenticate.new(id: id)
+      end
+    end
+
+    show
+  end
```

Comment:
This seems like a really odd way of doing authentication.  But I think our authentication is a bit odd across all of our apps at the moment.  This is fine for now.